c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -3676,6 +3681,133 @@ bool response_started(JSObject *self) {
+using TimerArgumentsVector = std::vector<JS::Heap<JS::Value>>;
+
+class Timer {
+public:
+  uint32_t id;
+  JS::Heap<JSObject *> callback;
+  TimerArgumentsVector arguments;
+  uint32_t delay;
+  system_clock::time_point deadline;
+  bool repeat;
+
+  Timer(uint32_t id, HandleObject callback, uint32_t delay, system_clock::time_point deadline,
+        JS::HandleValueVector args, bool repeat)
+      : id(id), callback(callback), delay(delay), deadline(deadline), repeat(repeat) {
+    deadline = system_clock::now() + system_clock::duration(delay * 1000);
+    arguments.reserve(args.length());
+    for (auto arg : args) {
+      arguments.push_back(JS::Heap(arg));
+    }
+  }
+
+  friend bool operator<(Timer const &left, Timer const &right) {
+    return left.deadline < right.deadline;
+  }
+
+  void trace(JSTracer *trc) {
+    JS::TraceEdge(trc, &callback, "Timer callback");
+    for (auto &arg : arguments) {
+      JS::TraceEdge(trc, &arg, "Timer callback arguments");
+    }
+  }
+};
+
+class ScheduledTimers {
+public:
+  Timer *first() { return *timers.begin(); }
+
+private:
+  std::list<Timer *> timers;
+  static uint32_t next_id;
+
+  uint32_t add_timer(Timer *timer) {
+    auto iter = timers.begin();
+
+    for (; iter != timers.end(); iter++) {
+      if ((*iter)->deadline > timer->deadline) {
+        break;
+      }
+    }
+
+    auto inserted = (*timers.insert(iter, timer));
+    return inserted->id;
+  }
+
+  void repeat_first() {
+    Timer *timer = first();
+    timer->deadline = system_clock::now() + system_clock::duration(timer->delay * 1000);
+    timers.remove(timer);
+    add_timer(timer);
+  }
+
+public:
+  bool empty() { return timers.empty(); }
+
+  uint32_t add_timer(HandleObject callback, uint32_t delay, JS::HandleValueVector arguments,
+                     bool repeat) {
+    auto deadline = system_clock::now() + system_clock::duration(delay * 1000);
+    auto timer = new Timer(next_id++, callback, delay, deadline, arguments, repeat);
+    return add_timer(timer);
+  }
+
+  void remove_timer(uint32_t id) {
+    for (auto timer : timers) {
+      if (timer->id == id) {
+        timers.remove(timer);
+        break;
+      }
+    }
+  }
+
+  bool run_first_timer(JSContext *cx) {
+    Timer *timer = first();
+    MOZ_ASSERT(system_clock::now() > timer->deadline);
+    uint32_t id = timer->id;
+    RootedObject fun(cx, timer->callback);
+    JS::RootedVector<JS::Value> argv(cx);
+    if (!argv.initCapacity(timer->arguments.size())) {
+      JS_ReportOutOfMemory(cx);
+      return false;
+    }
+
+    for (size_t i = 0; i < timer->arguments.size(); i++) {
+      argv.infallibleAppend(timer->arguments[i]);
+    }
+
+    // The timer might be deleted by content while running, so we mustn't access it again here.
+    timer = nullptr;
+
+    RootedValue rval(cx);
+    if (!JS::Call(cx, JS::NullHandleValue, fun, argv, &rval)) {
+      return false;
+    }
+
+    // Repeat / remove the first timer if it's still the one we just ran.
```

Comment:
In what situations would the timer at the head of the queue no longer be the one we just ran? Would it be easier to pop the timer off of the front of the queue and take ownership of it for the duration of the processing? That way you wouldn't have to worry about whether or not something was pushed in the meantime.

Reply:
I believe this can happen if the timer being ran adds a new timer which has a deadline that is earlier than the timer being ran. I'll see if that is actually possible to occur within a JS application

Reply:
Ah I see: if this timer is set to repeat, then it moves the timer back in the queue rather than re-allocate the structure.